### PR TITLE
Top level templates

### PR DESCRIPTION
**Pull request by leoyvens**
See https://github.com/graphprotocol/graph-node/issues/1023. Both formats are supported, but not a mix of both in the same subgraph. If a subgraph tries to mix both, only the top-level templates are considered. The `RuntimeHostBuilder` is given the data and the logic to make the decision of which template format is in use. See commits for more details.

A current limitation is that a data source cannot spawn another in the same block that it was created, for now just detect and error in this case, but this should be fixed in a follow up. See https://github.com/graphprotocol/graph-node/issues/1105.

graph-cli PR https://github.com/graphprotocol/graph-cli/pull/319

### CODE REVIEW

File: core/src/subgraph/instance_manager.rs [hunk 1]
```diff
@@ -508,7 +501,17 @@ where
...
                 triggers,
             )
-            .map(move |block_state| (ctx, block_state, data_sources, runtime_hosts))
+            .and_then(move |block_state| {
+                match block_state.created_data_sources.len() > created_ds_count {
+                    false => Ok((ctx, block_state, data_sources, runtime_hosts)),
+                    true => Err(err_msg(
+                        "A dynamic data source, in the same block that it was created,
+                        attempted to create another dynamic data source.
+                        Let us know in graph-node issue #1105 that you hit this.",
```
Comment (Jannis) [inline core/src/subgraph/instance_manager.rs]: This is a great idea. Can we replace `Let us know in graph-node issue #1105 that you hit this` with `Let us know if you hit this on https://github.com/graphprotocol/graph-node/issues/1105`? That way people can click on the link in their terminals if they run their own Graph node.

File: graph/src/data/subgraph/mod.rs [hunk 2]
```diff
@@ -898,27 +900,51 @@ impl UnresolvedSubgraphManifest {
...
             data_sources,
+            templates,
         } = self;
 
+        match semver::Version::parse(&spec_version) {
+            Ok(ref ver) if *ver <= semver::Version::new(0, 0, 2) => {}
+            _ => {
+                return Box::new(future::err(format_err!(
+                    "This Graph Node only supports manifest spec versions <= 0.0.2,
+                but subgraph `{}` uses `{}`",
```
Comment (Jannis) [inline graph/src/data/subgraph/mod.rs]: Is the message formatting / indentation correct here? Looks like it'll print an error message with the second line way indented.

File: graph/src/data/subgraph/mod.rs [hunk 3]
```diff
@@ -898,27 +900,51 @@ impl UnresolvedSubgraphManifest {
...
+                    id,
+                    spec_version
+                ))) as Box<dyn Future<Item = _, Error = _> + Send>;
+            }
+        }
+
         // resolve each data set
```
Comment (Jannis) [inline graph/src/data/subgraph/mod.rs]: I'm just noticing now that we're using `data set` and `data_set` in this method. Fancy changing that to `data source` / `data_source` (optional for this review)?

**Review by Jannis — CHANGES_REQUESTED** [review]
Looks good, only a couple of small comments.

**Comment by leoyvens** [comment]
@Jannis addressed your comments.

File: graph/src/data/subgraph/mod.rs (hunk 2 above)
Comment (leoyvens) [reply graph/src/data/subgraph/mod.rs]: This should print as a single line with a single space in the line break, but I've aligned it nicely.

**Review by Jannis — APPROVED** [review]

File: core/src/subgraph/instance_manager.rs [hunk 4]
```diff
@@ -508,7 +501,18 @@ where
...
             )
-            .map(move |block_state| (ctx, block_state, data_sources, runtime_hosts))
+            .and_then(move |block_state| {
+                match block_state.created_data_sources.len() > created_ds_count {
+                    false => Ok((ctx, block_state, data_sources, runtime_hosts)),
+                    true => Err(err_msg(
+                        "A dynamic data source, in the same block that it was created,
+                        attempted to create another dynamic data source.
+                        To let us know that you are affected by this bug, please comment in
+                        https://github.com/graphprotocol/graph-node/issues/1105",
```
Comment (fordN) [inline core/src/subgraph/instance_manager.rs]: Nice, I like the approach of pushing users to a specific issue if they encounter the problem.

**Review by fordN — APPROVED** [review]

**Pull request merged**